README.md
# ipv6gen
IPv6 prefix generator

ipv6gen is tool which generates list of IPv6 prefixes of given length from certain prefix 
according to http://www.ietf.org/rfc/rfc3531.txt'>RFC 3531. (A Flexible Method for Managing the Assignment of 
Bits of an IPv6 Address Block)

This is intended as a helper script for either constructing IPv6 addressing scheme or allocating prefixes automatically. 
ipv6gen is structured into functions, which can be used in proprietary scripts.

ipv6gen is written in Perl.

ipv6gen and other related scripts are GPL licensed. 
If you take some code out of ipv6gen and use it in your project, it would be nice if you mention that it comes from ipv6gen.

How it works ?
  - input :
    - prefix from which we will be generating smaller prefixes
    - method of bit allocation (left, right, from the middle)
    - size of prefixes to be generated
  - output : list of prefixes according to input data

## Changelog

see Changelog.txt

Examples
ipv6gen

    Generating prefixes for subnets

    Say we want to generate /64 prefixes for application servers from
    2001:1508:1003::/48, each /64 prefix for servers of given type.
    If method specifier is omited, allocation from the
    right will be used:

    worm[~/IPv6/ipv6gen]> ./ipv6gen.pl 2001:1508:1003::/48 64


    2001:1508:1003:0000::/64


    2001:1508:1003:0001::/64


    2001:1508:1003:0002::/64


    2001:1508:1003:0003::/64





    ...





    2001:1508:1003:FFFA::/64


    2001:1508:1003:FFFB::/64


    2001:1508:1003:FFFC::/64


    2001:1508:1003:FFFD::/64


    2001:1508:1003:FFFE::/64


    2001:1508:1003:FFFF::/64


    Splitting prefixes

    Or we just need to split /48 prefix into two prefixes:

    worm[~/IPv6/ipv6gen]> ./ipv6gen.pl 2001:1508:1003::/48 49


    2001:1508:1003:0000::/49


    2001:1508:1003:8000::/49


    This is actually example of allocation which is prone to errors. First byte
    after /48 boundary is 10000000 in binary which is 128 in decimal, which is 80
    in hexa.
    Flexible bit allocation

    Or we need to flexibly allocate prefixes which will be used for further
    allocations by starting allocating bits in the middle. In this example,
    debugging option -d will be used to see that the bits are really allocated
    this way:

    worm[~/IPv6/ipv6gen]> ./ipv6gen.pl -d -m 2001:1508:1003::/48 53


    binary addr : 001000000000000100010101000010000001000000000011


    binary addr : 001000000000000100010101000010000001000000000011(48) 


    delta 5


    zero filled : 00000


    converting 001000000000000100010101000010000001000000000011 # 00000


    2001:1508:1003:0000::/53


    ----- round # 0


    left: 0


    central_bit: 3


    middle: 00100


    converting 001000000000000100010101000010000001000000000011 # 00100


    2001:1508:1003:2000::/53


    ----- round # 1


    left: 1


    central_bit: 2


    middle: 01000


    converting 001000000000000100010101000010000001000000000011 # 01000


    2001:1508:1003:4000::/53


    middle: 01100


    converting 001000000000000100010101000010000001000000000011 # 01100


    2001:1508:1003:6000::/53


    ----- round # 2


    left: 0


    central_bit: 4


    middle: 00010


    converting 001000000000000100010101000010000001000000000011 # 00010


    2001:1508:1003:1000::/53


    middle: 00110


    converting 001000000000000100010101000010000001000000000011 # 00110


    2001:1508:1003:3000::/53


    middle: 01010


    converting 001000000000000100010101000010000001000000000011 # 01010


    2001:1508:1003:5000::/53


    middle: 01110


    converting 001000000000000100010101000010000001000000000011 # 01110


    2001:1508:1003:7000::/53





    ...


    Leaving gaps between generated prefixes

    ipv6gen can also leave out gaps between each two generated prefixes.
    This is usefull when designing allocation schemes which should be
    extendable. Such scheme is for example using RIPE. Currently it gives out
    /32 prefixes to LIRs. For each LIR, it leaves out 7 subsequent prefixes,
    which will be used in case LIR will need more address space. These
    subsequent prefixes can be aggregated up to to /29 prefix.

    For example, one of the prefixes
    allocated
    by IANA to RIPE NCC is
    2001:1400::/23 prefix. From this prefix, RIPE NCC further
    allocates /32 prefixes (in general) to LIRs. These prefixes are allocated
    using rightmost allocation. For each prefix, 7 subsequent prefixes are
    left out. So, RIPE could use ipv6gen for this task:

    simstim:ipv6gen $ ./ipv6gen.pl -s 8 -r 2001:1400::/23 32


    2001:1400::/32


    2001:1408::/32


    2001:1410::/32


    2001:1418::/32


    2001:1420::/32


    2001:1428::/32


    2001:1430::/32


    2001:1438::/32


    ...


    Compare this output with
    list of
    prefixes allocated by RIPE to LIRs to see that this is really the scheme
    RIPE uses.

check-overlap

    This script checks if one prefix is allocated from second prefix.

    But check-overlap script does more than that. It does 3 kinds of checks:

            IPv6 address format check
            It checks if IPv6 address format is correct (according to
            RFC3513, section 2.2.
            )
            XXX: implemented only in ipv6gen
            correctness of IPv6 prefix address
            Checks for bits with value 1 beyond prefixlen boundary.
            This check is also implemented in ipv6gen.
            check for overlapping prefixes
            Check if two prefixes overlap - i.e. second prefix was allocated
            from the first one.

    If one of the checks fails, following checks are not made:

    worm[~/IPv6/ipv6gen]> ./check-overlap.pl 2001:1508:1000:FFFF::/45 2001:1508:1000:FF00:/56


    ----  doing prefix checks first


    --- checking 2001:1508:1000:FFFF::/45


    error in prefix : bit 1 found beyond prefixlen boundary


    binary representation of the prefix:


    0010000000000001000101010000100000010000000000001111111111111111


    001000000000000100010101000010000001000000000 0001111111111111111


                                              /45    ^


    --- checking 2001:1508:1000:FF00:/56 ... OK


    For some prefixes it is not visible on the first sight,
    if they are overlapping:

    worm[~/IPv6/ipv6gen]&gt; ./check-overlap.pl 2001:1508:FFD0:40::/50 2001:1508:FFB0:C0::/51


    ----  doing prefix checks first


    --- checking 2001:1508:FFD0:40::/50 ... OK


    --- checking 2001:1508:FFB0:C0::/51 ... OK


    --- checking overlap ... prefixes do NOT overlap


    00100000000000010001010100001000111111111101000001 000000


    00100000000000010001010100001000111111111011000011 000000


                                                   /50


    ... but these are not.

    These are clear:

    worm[~/IPv6/ipv6gen]> ./check-overlap.pl 2001:1508::/32 2001:1508:FFB0:C0::/52


    ----  doing prefix checks first


    --- checking 2001:1508::/32 ... OK


    --- checking 2001:1508:FFB0:C0::/52 ... OK


    --- checking overlap ... OK (overlapping)


## Known bugs/TODO

- add support of compressed IPv6 address format according to RFC3513, section 2.2.
- make Perl code more clean
- make a Perl library for work with IPv6 prefixes (?)
- better man page
- make ipv6gen available as cgi script


Changelog.txt

1.0 - fixed bug in the algorithm allocating bits from the middle when
      the delta was only 1 bit
0.9 - fixed bug centermost algorithm :
      at the very first round the bit following the center for the even 
      number of bits was not picked
      - from Muggle L <muggle19355--at--gmail.com>
0.8 - added checking of semantical correctness of input prefix
      (if there are non-zero bits after prefix length boundary,
      prefix is considered bad and ipv6gen exits with error.)
0.7 - added stepping between generated prefixes as input parameter
0.6 - fixed RFC 3513 non-compliancy bug (ipv6gen appended :: after
      prefixes longer than 112 bits)
0.5 - fixed zero padding bug - ipv6gen incorrectly generated prefixes
      not aligned to byte boundary. e.g. /64 prefixes were incorrectly
      generated from /49 prefixes)
0.4 - fixed bug when generating bits for middle allocation and 
      delta with even number of bits 
0.3 - check for format of IPv6 prefixes
0.2 - implemented allocation from the middle
0.1 - only allocation from the left/right is implemented


check-overlap.pl
#!/usr/bin/perl
#
# check if two IPv6 prefixes overlap correctly
#
# functions for conversions betweeen hex and binary are
# based on script hex2bin.pl by Stas Bekman <stas@stason.org>
#
# Vladimir Kotal <vlada@devnull.cz>, 2005
#

use Getopt::Std;
use strict;

$0 =~ s|^.*/||;        # cut the basename of the command

# opt vars
my $quiet = 0;
my $debug = 0;

# global variables
my $error = 0;

# Define the separation character between each binary number
my $char = " ";

# print help and exit
sub help () {
  print "usage: $0 <bigger_prefix> <smaller_prefix>\n";
  exit(1);
}

sub debug () {
  print @_ if ($debug > 0);

}

# convert one char in hex to binary
sub hex2bin_c {
    my ($digit) = shift;
    return unpack("B4", pack("H", $digit));
}


# convert string with hex representation to binary
sub hex2bin {
  my $in = shift;
  my $out;

  my @digits = split //, $in;
  for (@digits) {
    $out = $out . hex2bin_c($_);
  }
  return $out;
}


# XXX
sub addr2bin {
  my $prefix = shift;
  my @sects = split(/:/, $prefix);
  my $s;
  my $tmp;
  my $addr_bin = "";

  for $s (@sects) {
      $tmp = &hex2bin($s);
      $addr_bin = $addr_bin . $tmp;
  }
  return $addr_bin;
}


#
# print binary representation of prefix split to 2 sections
#
# arg1: binary representation of prefix
# arg2: length
#
sub print_split {
  my $bits = shift;
  my $len= shift;

  &debug("$bits\n");
  print   substr($bits, 0, $len) 
	. " " 
	. substr($bits, $len) 
	. "\n";
}

# XXX
sub print_boundary_mark () {
  my $idx = shift;
  my $len = shift;

  my $i;

  for ($i = 0; $i < $len - 3; $i++) {
    print " ";
  }
  print "\/$len";
  if ($idx > 0) {
    for ($i = 0; $i < $idx + 1; $i++) {
      print " ";
    }
    print "^";
  }
  print "\n";
}

#
# check for 1 bits behind prefix length limit
#
# arg1: binary representation of address
# arg2: length of prefix
#
sub check_pfx_bin () {
  my $bits = shift;
  my $len = shift;
  my $idx;

  my $beyond_pfxlen = substr($bits, $len);
  if ($beyond_pfxlen =~ /1/) {
    if ($quiet != 1) {
      print "\nerror in prefix : bit 1 found beyond prefixlen boundary\n";
      print "binary representation of the prefix:\n";
      &print_split($bits, $len);

      $idx = index($beyond_pfxlen, "1"); 
      &print_boundary_mark($idx, $len);
    }
    $error++;
    return(1);
  }
}


#
# check if smaller prefix fits into bigger prefix
# 
# arg1: bigger IPv6 prefix 
# arg2: smaller IPv6 prefix
#
sub check_overlap {
  my $prefix1 = shift;
  my $prefix2 = shift;

  my $big_pfx_part;
  my $big_pfx_len;
  my $small_pfx_part;
  my $small_pfx_len;

  ($big_pfx_part, $big_pfx_len) = split(/\//, $prefix1);
  ($small_pfx_part, $small_pfx_len) = split(/\//, $prefix2);

  my $big_bits = &addr2bin($big_pfx_part);
  my $small_bits = &addr2bin($small_pfx_part);

  # smaller prefixlen makes bigger prefix
  if ($big_pfx_len > $small_pfx_len) {
    print "bigger prefix should be specified first\n";
    return(1);
  }

  print "--- checking overlap" if ($quiet == 0); 

  my $f = substr($big_bits, 0, $big_pfx_len);
  my $s = substr($small_bits, 0, $big_pfx_len);
  if ($f cmp $s) {
    print " ... prefixes do NOT overlap\n" if ($quiet == 0);
    # print "$f\n$s\n";
    &debug("$prefix1\n");
    &print_split($big_bits, $big_pfx_len);
    &debug("$prefix2\n");
    &print_split($small_bits, $big_pfx_len);
    &print_boundary_mark(0, $big_pfx_len);
    $error++;
    return(1);
  }
  print " ... OK (overlapping)\n" if ($quiet == 0);
}


#
# check prefix
# this is really just a wrapper to check_pfx_bin()
#
# arg1: prefix
#
sub check_pfx () {
  my $prefix = shift;
  my $pfx_part;
  my $pfx_len;

  ($pfx_part, $pfx_len) = split(/\//, $prefix);

  my $pfx_bits = &addr2bin($pfx_part);
  
  print "--- checking $prefix" if ($quiet == 0); 
  if (!&check_pfx_bin($pfx_bits, $pfx_len)) {
    print " ... OK\n" if ($quiet == 0);
  }
}


#
# check format of prefix
# arg1: prefix
# XXX
# 
sub check_fmt() {
  my $prefix = shift;

  return(0);
}

# 
# MAIN
#
my $opts = {};
&getopts('qdh', $opts);
&help if exists $opts->{"h"};
$debug = 1 if exists $opts->{"d"};
$quiet = 1 if exists $opts->{"q"};

&help if (scalar(@ARGV) != 2);


# check IPv6 prefix format
# XXX check if IPv6 address has the right format
if (&check_fmt($ARGV[0]) || &check_fmt($ARGV[1])) {
  print "bad prefix format\n";
  exit(1);
}

# run the checks
print "----  doing prefix checks first\n" if ($quiet == 0); 
&check_pfx($ARGV[0]);
&check_pfx($ARGV[1]);

&check_overlap($ARGV[0], $ARGV[1]) if ($error == 0);

exit (1) if ($error);


ipv6gen.pl
#!/usr/bin/perl
#
# IPv6 prefix generator
#   - generates list of prefixes of certain size from given prefix
#     with one of 3 methods for bit allocation (according to
#     RFC3531)
#
#
# Vladimir Kotal <vlada@devnull.cz>, 2005
#

use Getopt::Std;
use POSIX qw(ceil floor);
use strict;

$0 =~ s|^.*/||;        # cut the basename of the command

$|=1;

# variables
my $debug = 0;		# 1 - print debug messages
my $method = "r";	# strategy for generating prefixes, default : right
my $method_set = 0;	# indicate if strategy was set
my $step = 1;		# default step between generated prefixes
my $version = "0.9";

#
# print help and exit
#
sub help ()
{
  print "$0 (v$version) - generates IPv6 prefixes from given \n";
  print "                    IPv6 prefix pool using given strategy\n\n";
  print "usage: ipv6gen [switch|strategy] <pool> <prefix length>\n";
  print "\nstrategies:\n";
  print "-l\t\tleft to right\n";
  print "-r\t\tright to left\n";
  print "-m\t\tfrom the middle out\n";
  print "\nswitches:\n";
  print "-d\t\tdisplay debug messages\n";
  print "-s <num>\t\tstep between prefixes\n";
  print "\n";

  exit(1);
}

sub debug () {
  print @_ if ($debug > 0);

}

#
# pad string from left with zeroes
#
# arg1: string to be padded
# arg2: length of resulting string
# arg3: 1 - pad from left, 0 - pad from right
#
sub zero_pad () {
  my $str = shift;
  my $len = shift;
  my $left = shift;
  
  my $i;

  my $strlen = length($str);
  for ($i = 0; $i < $len - $strlen; $i++) {
    if ($left) {
      $str = "0" . $str;
    } else {
      $str = $str . "0"; 
    }
  }
  return $str;
}

# convert dec to bin 
# and output binary number w/out leading zeroes
sub dec2bin_old() { 
  my $arg = shift @_;
  my $str = unpack('B32', pack('C', $arg)) . "";

  # strip leading zeroes
  if ($str !~ /00000000/) {
    $str =~ s/^0+// 
  } else {
    $str = "0";
  }
  return $str;
}

sub dec2bin {
  my $str = unpack("B32", pack("N", shift));
  $str =~ s/^0+(?=\d)//;   # otherwise you'll get leading zeros
  return $str;
}

# convert one char in hex to binary
sub hex2bin_c {
    my ($digit) = shift;
    return unpack("B4", pack("H", $digit));
}

# convert string with hex representation to binary
sub hex2bin {
  my $in = shift;
  my $out;

  my @digits = split //, $in;
  for (@digits) {
    $out = $out . hex2bin_c($_);
  }
  return $out;
}

# convert one char from binary to hex
# input char must be 4 bits long
sub bin2hex_c {
  my $binary = shift;

  my $str = unpack("H", pack("B4", $binary));
  $str =~ tr/a-f/A-F/;
  return $str;
}

# convert binary representation of IPv6 address to colon format
#
# arg1: binary represenatation of IPv6 addr
# 
# NOTE: this function does not output valid IPv6 address
#
sub binaddr_to_dots {
  my $bin = shift;

  my $out = "";
  my $first = 1;
  my $converted = 0; # number of bits converted
  my $cbd = 0; # number of chars beyond :

  my $conv;
  my $lastconv;
  # &debug("bin : $bin\n");

  while (length($bin) > 0) {
    $conv = substr($bin, 0, 4);
    # print "conv : $conv\n";
    $lastconv = bin2hex_c($conv);
    $out = $out . $lastconv;
    $converted += length($conv);
    $cbd++;
    if ((!$first) && !($converted % 16) && (length($bin) != 4)) {
      # $binlen = length($bin);
      # &debug("adding ':' [$binlen]\n");
      $out = $out . ":";
      $cbd = 0;
    }
    $first = 0;
    $bin = substr($bin, 4, length($bin) - 4);
  }
  # print "converted = $converted\n";
  # pad with null till next : boundary
  if ($cbd > 0) {
    $out = $out . substr("0000", 0, 4 - $cbd);
  }
  return $out;
}

#
# convert IPv6 address to binary
# XXX prefixlen ?
#
# arg1: XXX
#
sub addr2bin {
  my $prefix = shift;

  my @sects = split(/:/, $prefix);
  my $s;
  my $tmp;
  my $addr_bin;

  for $s (@sects) {
      # pad to zeroes from the left according to RFC 3513, section 2.3
      # (Internet Protocol Version 6 (IPv6) Addressing Architecture)
      $s = &zero_pad($s, 4, 1);
      $tmp = &hex2bin($s);
      $addr_bin = $addr_bin . $tmp;
  }
  return $addr_bin;
}

#
# check prefix format and print it if it passes the check
#
# arg1: IPv6 prefix (with prefixlen)
#
sub print_pfx () {
  my $pfx = shift;

  if (&ipv6pfx_check($pfx)) {
    print "$pfx\n";
  } else {
    print "bad prefix format: $pfx\n";
    print "#### please report this bug\n";
    exit(1);
  }
}

#
# check if IPv6 address is valid
#
# idea and parts of the regexp by Glynn Beeken via regexlib.com
#
# arg1: prefix (in format addr/prefixlen)
# return: -1 on error, 1 otherwise
# 
sub ipv6pfx_check () {
  my $prefix = shift;

  my $addr; 
  my $pfx_len;

  ($addr, $pfx_len) = split(/\//, $prefix);
  &debug("checking $prefix = $addr / $pfx_len\n");
  return (-1) if ($pfx_len > 128);

  return (1) if ($addr =~ /([0-9A-Fa-f]{1,4}:){7}[0-9A-Fa-f]{1,4}/);
  return (1) if ($addr =~ /[0-9A-Fa-f]{1,4}(:[0-9A-Fa-f]{1,4}){5}::[0-9A-Fa-f]{1,4}/);
  return (1) if ($addr =~ /[0-9A-Fa-f]{1,4}(:[0-9A-Fa-f]{1,4}){4}::[0-9A-Fa-f]{1,4}(:[0-9A-Fa-f]{1,4}){0,1}/);
  return (1) if ($addr =~ /[0-9A-Fa-f]{1,4}(:[0-9A-Fa-f]{1,4}){3}::[0-9A-Fa-f]{1,4}(:[0-9A-Fa-f]{1,4}){0,2}/);
  return (1) if ($addr =~ /[0-9A-Fa-f]{1,4}(:[0-9A-Fa-f]{1,4}){2}::[0-9A-Fa-f]{1,4}(:[0-9A-Fa-f]{1,4}){0,3}/);
  return (1) if ($addr =~ /[0-9A-Fa-f]{1,4}(:[0-9A-Fa-f]{1,4}::[0-9A-Fa-f]{1,4}(:[0-9A-Fa-f]{1,4}){0,4})/);
  return (1) if ($addr =~ /[0-9A-Fa-f]{1,4}::[0-9A-Fa-f]{1,4}(:[0-9A-Fa-f]{1,4}){0,5}/);
  return (1) if ($addr =~ /[0-9A-Fa-f]{1,4}(:[0-9A-Fa-f]{1,4}){0,6}::/);
  return (1) if ($addr =~ /::[0-9A-Fa-f]{1,4}(:[0-9A-Fa-f]{1,4}){0,6}/);
  return (1) if ($addr =~ /::/);

  return(-1);
}

#
# print binary representation of prefix split to 2 sections
#
# arg1: binary representation of prefix
# arg2: length
# NOTE: taken from check-overlap.pl
#
sub print_split {
  my $bits = shift;
  my $len= shift;

  &debug("$bits\n");
  print   substr($bits, 0, $len) 
	. " " 
	. substr($bits, $len) 
	. "\n";
}

# 
# print boundary mark
# arg1: index
# arg2: length
# NOTE: taken from check-overlap.pl
#
sub print_boundary_mark () {
  my $idx = shift;
  my $len = shift;

  my $i;

  for ($i = 0; $i < $len - 3; $i++) {
    print " ";
  }
  print "\/$len";
  if ($idx > 0) {
    for ($i = 0; $i < $idx + 1; $i++) {
      print " ";
    }
    print "^";
  }
  print "\n";
}

#
# check for 1 bits behind prefix length limit
#
# arg1: binary representation of address
# arg2: length of prefix
# NOTE: taken from check-overlap.pl
#
sub check_pfx_bin () {
  my $bits = shift;
  my $len = shift;
  my $idx;

  my $beyond_pfxlen = substr($bits, $len);
  if ($beyond_pfxlen =~ /1/) {
    print("\nerror in prefix : bit 1 found beyond prefixlen boundary\n");
    print("binary representation of the prefix:\n");
    &print_split($bits, $len);

    $idx = index($beyond_pfxlen, "1"); 
    &print_boundary_mark($idx, $len);
    exit(1);
  }
}


# convert binary representation of prefix and print it on STDOUT 
sub print_binpfx () {
  my $pfx = shift;
  my $append = shift;
  my $glen = shift;

  &debug("converting $pfx $append\n");
  my $addr = &binaddr_to_dots($pfx . $append);
  # do not append :: to prefixes with length bigger than 112
  # in order to be compliant with RFC 3513, section 2.2
  # NOTE: this would have to be changed with support of compressed format
  if ($glen <= 112) {
    $addr = $addr . "::";
  }
  &print_pfx("$addr/$glen");
}

#
# generate list of prefixes by allocating bits from the middle
# NOTE: this function actually prints prefixes to STDOUT
#
# arg1: binary representation of prefix pool addr
# arg2: length of prefix pool
# arg3: length of prefixes to be generared from pool
#
sub generate_m () {
  my $addr_bin_strip = shift;
  my $plen = shift;	
  my $glen = shift;

  my $i;
  my $rnd;
  my $str;
  my $addr;

  my $left = 0;

  my $delta = $glen - $plen;
  &debug("delta $delta\n");
  # number of rounds is equal to number of bits to be filled
  my $limit = $delta;
  my $rnd_limit = 0;

  # position of first bit with value 1
  my $central_bit = ceil($delta / 2);  

  # zero-filled str is speciall case
  $str = &zero_pad($str, $delta, 1);
  &debug("zero filled : $str\n");
  &print_binpfx($addr_bin_strip, $str, $glen);
  
  if (($delta % 2) == 0) {
    $left = 1;
    # Even number of bits, pick the bit following the center
    # RFC3531 Section 3.3
    # muggle19355@gmail.com
    $central_bit++; 
  }
  
  for ($rnd = 0; $rnd < $limit; $rnd++) {
    $rnd_limit=2 ** $rnd; # number of passes in each round

    # special case for even number of bits at the very first round
    if ($rnd_limit == 1)
    {
        if ($left == 1 && ($delta % 2) == 0)
        {
            $left=0;
        }
    }

    &debug("----- round \# $rnd\n");
    &debug("left: $left\n");
    &debug("central_bit: $central_bit\n");
    for ($i = 0; $i < $rnd_limit; $i++) {
      $str = &dec2bin($i);
      #&debug("dec2bin: $i = $str\n");
   
      my $tmp = "";
      # $str = &zero_pad($str, $central_bit, 1); 
      if ($left == 1) {
	$tmp = &zero_pad($tmp, $central_bit - 1, 1);
        $tmp = $tmp . "1";
        # &debug("tmp: $tmp\n");
        $str = reverse $str; 
	$str = $tmp . $str;
      } else {
	# This will take care of the delta == 1 case
	if ($i == 0) {
	  $str = "1";
	} else {
	  $str = $str . "1";
	}
        $str = &zero_pad($str, $central_bit, 1);
      }
      $str = &zero_pad($str, $delta, 0);
      &debug("middle: $str\n");
      &print_binpfx($addr_bin_strip, $str, $glen);
    }
    $left = ($left + 1) % 2;
    if ($left) {
      $central_bit = $central_bit - $rnd -1;
    } else {
      $central_bit = $central_bit + $rnd +1;
    }
  }
}

#
# generate list of prefixes by allocating bits from the left/right
# NOTE: this function actually prints prefixes to STDOUT
#
# arg1: binary representation of prefix pool addr
# arg2: length of prefix pool
# arg3: length of prefixes to be generared from pool
# arg4: 1 - generate from left, 0 - right
#
sub generate_rl () {
  my $addr_bin_strip = shift;
  my $plen = shift;	
  my $glen = shift;
  my $reverse = shift;

  my $i;
  my $str;
  my $addr;

  my $delta = $glen - $plen;
  &debug("delta $delta\n");
  my $limit = 2 ** $delta;
  die "step bigger than limit ($step > $limit)" if ($step > $limit);

  &debug("will generate " . $limit/$step . " prefixes\n");
  for ($i = 0; $i < $limit; $i+=$step) {
    $str = &dec2bin($i);
    # my $debug_str = reverse $str if ($reverse);
    # &debug("generated bits: $debug_str\n");
    $str = &zero_pad($str, $delta, 1);
    $str = reverse $str if ($reverse);

    &print_binpfx($addr_bin_strip, $str, $glen);
  }
}

#
# generate IPv6 prefixes from IPv6 prefix pool using given strategy
# arg1: prefix to generate from
# arg2: length of prefixes to be generated
# arg3: method for
#
sub generate () {
  my $prefix = shift; 	# prefix pool
  my $plen = shift; 	# length of prefix pool
  my $glen = shift; 	# length of prefixes to be generated
  my $method = shift;

  my $addr_bin = &addr2bin($prefix);
  &check_pfx_bin($addr_bin, $plen);
  
  &debug("binary addr : $addr_bin (not padded/cropped)\n");
  # pad the prefix to $plen, otherwise we would get wrong results
  $addr_bin = &zero_pad($addr_bin, $plen, 0);
  my $addr_bin_strip = substr($addr_bin, 0, $plen);
  &debug("binary addr : " . $addr_bin_strip . "($plen) \n");

  # call proper generating function
  if ($method =~ /r/) {
    &generate_rl($addr_bin_strip, $plen, $glen, 0);
  } elsif ($method =~ /l/) {
    &generate_rl($addr_bin_strip, $plen, $glen, 1);
  } elsif ($method =~ /m/) {
    &generate_m($addr_bin_strip, $plen, $glen);
  }
}

#
# set strategy for generating prefixes
#
sub set_method () {
  $method = shift @_;

  die "only single step is possible for middle bit allocation" 
    if (($step > 1) and ($method =~ /^m$/));
  &debug("setting method $method\n");

  &help if ($method_set == 1);
}

# --------------------------------------------------------------------------
# MAIN
# --------------------------------------------------------------------------

my $opts = {};
&getopts('s:dhmlr', $opts);
&help if exists $opts->{"h"};
$debug = 1 if exists $opts->{"d"};
$step = $opts->{"s"} if exists $opts->{"s"};

&help if (scalar(@ARGV) != 2);

&set_method("l") if exists $opts->{"l"};
&set_method("r") if exists $opts->{"r"};
&set_method("m") if exists $opts->{"m"};

if (&ipv6pfx_check($ARGV[0]) < 0) {
  print "bad prefix $ARGV[0]\n";
  exit(1);
}

my $addr;
my $pfx_len;

($addr, $pfx_len) = split(/\//, $ARGV[0]);

my $genlen = $ARGV[1];
if ($genlen !~ /^[0-9]+$/) {
  print "length of prefix to be generated is not number: $genlen\n";
  exit(1);
} 

if ($genlen < $pfx_len) {
 print "cannot generate \/$genlen prefixes from \/$pfx_len prefix\n";
 exit(1);
}

&generate($addr, $pfx_len, $genlen, $method);

# EOF
